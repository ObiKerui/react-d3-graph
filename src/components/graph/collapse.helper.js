/**
Developer notes - collapsing nodes and maintaining state on links matrix.

User interaction flow
1. User clicks node
2. All leaf connections of that node are not rendered anymore
3. User clicks on same node
4. All leaf connections of that node are rendered

rd3g calculations
1. User clicks node
2. Compute leaf connections for clicked node (targetNode)
3. Update connections matrix (based on 2.)
4. Update d3Links array with toggled connections (based on 2.)
*/

/**
 * Calculates degree (in an out) of some provided node.
 * @param {string|number} nodeId - the id of the node whom degree we want to compute.
 * @param {Object.<string, Object>} linksMatrix - an object containing a matrix of connections of the graph, for each nodeId,
 * there is an object that maps adjacent nodes ids (string) and their values (number).
 * @returns {Object.<string, number>} returns object containing in and out degree of the node:
 * - inDegree: number
 * - outDegree: number
 */
function computeNodeDegree(nodeId, linksMatrix = {}) {
    return Object.keys(linksMatrix).reduce(
        (acc, source) => {
            if (!linksMatrix[source]) {
                return acc;
            }

            const currentNodeConnections = Object.keys(linksMatrix[source]);

            return currentNodeConnections.reduce((_acc, target) => {
                if (nodeId === source) {
                    return {
                        ..._acc,
                        outDegree: _acc.outDegree + linksMatrix[source][target]
                    };
                }

                if (nodeId === target) {
                    return {
                        ..._acc,
                        inDegree: _acc.inDegree + linksMatrix[source][target]
                    };
                }

                return _acc;
            }, acc);
        },
        {
            inDegree: 0,
            outDegree: 0
        }
    );
}

/**
 * Given in and out degree tells whether degrees indicate a leaf or non leaf scenario.
 * @param {number} inDegree - in degree.
 * @param {number} outDegree - out degree.
 * @returns {boolean} whether given degrees represent a leaf node or not.
 */
function _isLeaf(inDegree, outDegree) {
    return inDegree <= 1 && outDegree <= 1;
}

/**
 * Given a node id we want to calculate the list of leaf connections
 * @param {string} rootNodeId - node who's leafs we want to calculate.
 * @param {Object.<string, Object>} linksMatrix - an object containing a matrix of connections of the graph, for each nodeId,
 * there is an object that maps adjacent nodes ids (string) and their values (number).
 * @returns {Array.<Object.<string, string>>} a list of leaf connections.
 * What is a leaf connection? A leaf connection is a link between some node A and other node B
 * where A has id equal to rootNodeId and B has inDegree 1 and outDegree 0 (or outDegree 1 but the connection is with A).
 */
function getTargetLeafConnections(rootNodeId, linksMatrix = {}) {
    const rootConnectionsNodesIds = Object.keys(linksMatrix[rootNodeId]);

    return rootConnectionsNodesIds.reduce((leafConnections, target) => {
        const { inDegree, outDegree } = computeNodeDegree(target, linksMatrix);

        if (_isLeaf(inDegree, outDegree)) {
            leafConnections.push({
                source: rootNodeId,
                target
            });
        }

        return leafConnections;
    }, []);
}

/**
 * Update matrix given array of connections to toggle.
 * @param {Object.<string, Object>} linksMatrix - an object containing a matrix of connections of the graph, for each nodeId,
 * there is an object that maps adjacent nodes ids (string) and their values (number).
 * @param {Array.<Object.<string, string>>} connections - connections to toggle on matrix.
 * @param {boolean} directed - tells whether linksMatrix represents a directed graph or not.
 * @returns {Object.<string, Object>} updated linksMatrix
 */
function toggleLinksMatrixConnections(linksMatrix, connections, directed) {
    return connections.reduce(
        (newMatrix, link) => {
            if (!newMatrix[link.source]) {
                newMatrix[link.source] = {};
            }

            if (!newMatrix[link.source][link.target]) {
                newMatrix[link.source][link.target] = 0;
            }

            newMatrix[link.source][link.target] = newMatrix[link.source][link.target] === 0 ? 1 : 0;

            if (!directed) {
                newMatrix[link.target][link.source] = newMatrix[link.source][link.target];
            }

            return newMatrix;
        },
        { ...linksMatrix }
    );
}

/**
 * Updates d3Links by toggling given connections
 * @param {Array.<Object>} d3Links - An array containing all the d3 links.
 * @param {Array.<Object.<string, string>>} connectionMatrix - connections to toggle.
 * @returns {Array.<Object>} updated d3Links.
 */
function toggleLinksConnections(d3Links, connectionMatrix) {
    return d3Links.map(d3Link => {
        const { source, target } = d3Link;
        // TODO: Confirm whether fallbacks are needed in this situation
        const sourceId = source.id || source;
        const targetId = target.id || target;

        // TODO: Improve this code
        if (!(connectionMatrix && connectionMatrix[sourceId] && connectionMatrix[sourceId][targetId])) {
            return { ...d3Link, isHidden: true };
        }

        if (connectionMatrix[sourceId][targetId] === 0) {
            return { ...d3Link, isHidden: true };
        } else {
            return { ...d3Link, isHidden: false };
        }
    });
}

export { computeNodeDegree, getTargetLeafConnections, toggleLinksMatrixConnections, toggleLinksConnections };
