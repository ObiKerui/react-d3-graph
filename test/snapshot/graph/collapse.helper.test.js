import * as collapseHelper from '../../../src/components/graph/collapse.helper';
import * as graphData from '../../../sandbox/data/default';
import { initializeGraphState } from '../../../src/components/graph/graph.helper';

describe('Collapse Helper', () => {
    let directedState;
    let defaultState;

    beforeAll(() => {
        directedState = initializeGraphState({ data: graphData, id: 'id', config: { directed: true } });
        defaultState = initializeGraphState({ data: graphData, id: 'id' });
    });

    describe('#computeNodeDegree', () => {
        describe('when graph is directed', () => {
            test('should calculate correct node degree for "Androsynth"', () => {
                const nodeId = 'Androsynth';

                expect(collapseHelper.computeNodeDegree(nodeId, directedState.links)).toEqual({
                    inDegree: 0,
                    outDegree: 7
                });
            });

            test('should calculate correct node degree for "Podship"', () => {
                const nodeId = 'Podship';

                expect(collapseHelper.computeNodeDegree(nodeId, directedState.links)).toEqual({
                    inDegree: 1,
                    outDegree: 0
                });
            });

            test('should calculate correct node degree for "Mycon"', () => {
                const nodeId = 'Mycon';

                expect(collapseHelper.computeNodeDegree(nodeId, directedState.links)).toEqual({
                    inDegree: 2,
                    outDegree: 3
                });
            });
        });

        describe('when graph is not directed', () => {
            test('should calculate correct node degree for "Androsynth"', () => {
                const nodeId = 'Androsynth';

                expect(collapseHelper.computeNodeDegree(nodeId, defaultState.links)).toEqual({
                    inDegree: 7,
                    outDegree: 7
                });
            });

            test('should calculate correct node degree for "Podship"', () => {
                const nodeId = 'Podship';

                expect(collapseHelper.computeNodeDegree(nodeId, defaultState.links)).toEqual({
                    inDegree: 1,
                    outDegree: 1
                });
            });

            test('should calculate correct node degree for "Mycon"', () => {
                const nodeId = 'Mycon';

                expect(collapseHelper.computeNodeDegree(nodeId, defaultState.links)).toEqual({
                    inDegree: 5,
                    outDegree: 5
                });
            });
        });
    });

    describe('#getTargetLeafConnections', () => {
        describe('when graph is directed', () => {
            test('should return expected leaf connections for node "Mycon"', () => {
                const nodeId = 'Mycon';

                expect(collapseHelper.getTargetLeafConnections(nodeId, directedState.links)).toEqual([
                    { source: 'Mycon', target: 'Podship' }
                ]);
            });

            test('should return expected leaf connections for node "VUX"', () => {
                const nodeId = 'VUX';

                expect(collapseHelper.getTargetLeafConnections(nodeId, directedState.links)).toEqual([
                    { source: 'VUX', target: 'Intruder' }
                ]);
            });

            test('should return expected leaf connections for node "Eluder"', () => {
                const nodeId = 'Eluder';

                expect(collapseHelper.getTargetLeafConnections(nodeId, directedState.links)).toEqual([]);
            });
        });

        describe('when graph is not directed', () => {
            test('should calculate correct node degree for "Androsynth"', () => {
                const nodeId = 'Mycon';

                expect(collapseHelper.getTargetLeafConnections(nodeId, defaultState.links)).toEqual([
                    { source: 'Mycon', target: 'Podship' }
                ]);
            });

            test('should return expected leaf connections for node "VUX"', () => {
                const nodeId = 'VUX';

                expect(collapseHelper.getTargetLeafConnections(nodeId, defaultState.links)).toEqual([
                    { source: 'VUX', target: 'Intruder' }
                ]);
            });

            test('should return expected leaf connections for node "Eluder"', () => {
                const nodeId = 'Eluder';

                expect(collapseHelper.getTargetLeafConnections(nodeId, defaultState.links)).toEqual([]);
            });
        });
    });

    describe('#toggleLinksMatrixConnections', () => {
        describe('when graph is directed', () => {
            test('should properly toggle passed connections in links matrix', () => {
                const connections = [
                    { source: 'Androsynth', target: 'Chenjesu' },
                    { source: 'Androsynth', target: 'Ilwrath' }
                ];

                expect(
                    collapseHelper.toggleLinksMatrixConnections(directedState.links, connections, true)
                ).toMatchSnapshot();
            });
        });

        describe('when graph is not directed', () => {
            test('should properly toggle passed connections in links matrix', () => {
                const connections = [
                    { source: 'Androsynth', target: 'Chenjesu' },
                    { source: 'Androsynth', target: 'Ilwrath' }
                ];

                expect(
                    collapseHelper.toggleLinksMatrixConnections(defaultState.links, connections, false)
                ).toMatchSnapshot();
            });
        });
    });
});
