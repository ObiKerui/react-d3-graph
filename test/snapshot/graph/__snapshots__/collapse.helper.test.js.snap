// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse Helper #toggleLinksMatrixConnections when graph is directed should properly toggle passed connections in links matrix 1`] = `
Object {
  "Androsynth": Object {
    "Chenjesu": 0,
    "Guardian": 1,
    "Ilwrath": 0,
    "Mycon": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Avenger": Object {},
  "Broodhmome": Object {},
  "Chenjesu": Object {
    "Broodhmome": 1,
    "Mycon": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Drone": Object {},
  "Eluder": Object {},
  "Guardian": Object {},
  "Ilwrath": Object {
    "Avenger": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Intruder": Object {},
  "Mycon": Object {
    "Podship": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Podship": Object {},
  "Spathi": Object {
    "Eluder": 1,
  },
  "Umgah": Object {
    "Drone": 1,
    "VUX": 1,
  },
  "VUX": Object {
    "Intruder": 1,
  },
}
`;

exports[`Collapse Helper #toggleLinksMatrixConnections when graph is not directed should properly toggle passed connections in links matrix 1`] = `
Object {
  "Androsynth": Object {
    "Chenjesu": 0,
    "Guardian": 1,
    "Ilwrath": 0,
    "Mycon": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Avenger": Object {
    "Ilwrath": 1,
  },
  "Broodhmome": Object {
    "Chenjesu": 1,
  },
  "Chenjesu": Object {
    "Androsynth": 0,
    "Broodhmome": 1,
    "Mycon": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Drone": Object {
    "Umgah": 1,
  },
  "Eluder": Object {
    "Spathi": 1,
  },
  "Guardian": Object {
    "Androsynth": 1,
  },
  "Ilwrath": Object {
    "Androsynth": 0,
    "Avenger": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Intruder": Object {
    "VUX": 1,
  },
  "Mycon": Object {
    "Androsynth": 1,
    "Chenjesu": 1,
    "Podship": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Podship": Object {
    "Mycon": 1,
  },
  "Spathi": Object {
    "Androsynth": 1,
    "Chenjesu": 1,
    "Eluder": 1,
    "Ilwrath": 1,
  },
  "Umgah": Object {
    "Androsynth": 1,
    "Chenjesu": 1,
    "Drone": 1,
    "Ilwrath": 1,
    "Mycon": 1,
    "VUX": 1,
  },
  "VUX": Object {
    "Androsynth": 1,
    "Chenjesu": 1,
    "Ilwrath": 1,
    "Intruder": 1,
    "Mycon": 1,
    "Umgah": 1,
  },
}
`;
